docs/src/proposals/sbf-program-abi-v2.md

```diff
@@ -0,0 +1,125 @@
+  - `InvocationDepth = 7`: `u16`
+  - `InvocationDepthMax = 8`: `u16`
+  - `InvocationStack = 9`: `[InstructionContext; InvocationDepthMax]`
+- InstructionContext metadata regions:
+  - Read-only unless last invocation frame (for CPI)
+  - `InstructionData = 10`: `[u8]`
+  - `NumberOfAccountsInInstruction = 11`: `u16`
+  - `NumberOfProgramsInInstruction = 12`: `u16`
+  - `InstructionAccountIndices = 13`: `[u32; NumberOfAccountsInInstruction]`
+  - `AccountIsSigner = 14`: `[bool; NumberOfAccountsInInstruction]`
+  - `AccountIsWritable = 15`: `[bool; NumberOfAccountsInInstruction]`
+- AccountData regions:
+  - One mapped memory region for each account in the instruction
+  - But virtual and physical addresses stay the same throughout the transaction (useful for CPI)
+  - Read-only or read-write depends on the instruction context
+  - Content: `[u8]`
+
+#### Writable Attributes
+Alternative options:
+- Three regions per account: Read-only metadata, writable metadata, account data
+- All meta data in readonly region, plus duplicate for writable metadata in extra region
+- Two regions: Definitely read-only metadata and potentially writable metadata including readonly accounts and unmapped accounts
+
+#### Encoding Alignment
+Alternative options:
+- Tightly pack everything
+- Tightly pack attributes, pad values
+- Pad attributes and their values
+
+#### Host and Guest (VM) Address Space for Indirections / Pointers
+Alternative options:
+- Store guest pointers only, map host on demand
+- Translate on every context switch (invocation, exit, syscall)
+- Store both guest and host, but encrypt host pointers
+
+### Interface
+
+#### Instruction Structs
+TODO: Design a structured interface to deal with the account indices of different instructions.
+Replacing: `keyed_account_at_index(keyed_accounts, first_instruction_account + x)`
+
+#### Tests & Mockups
+TODO: Design a better way to create the mocked instruction accounts.
+Replacing: `create_keyed_accounts_unified(&keyed_account_tuples)`
+
+### Syscalls
+
+#### Cross Program Invocation (CPI)
+The runtime automatically allocates an empty invocation frame (mapped as writable) above the currently active one. If a program wants to call another one it simply writes directly into that empty invocation frame and configures it. Then the program only needs a syscall to trigger the execution and that syscall does not need to copy data forth and back between userspace and runtime anymore. It only has to validate the invocation frame, record the instruction and run the called program.
+
+#### Return Data Accessors
+In the old ABI return data was implemented by a setter and a getter syscall for copying data forth and back between userspace and runtime. But now, we can use shared memory in the transaction context instead. So, the two syscalls will be deprecated as there is no need to trigger anything in the runtime.
+
+#### Account Reallocation / Resizing
+TODO
```

Comment:
@Lichtso do you have any initial thoughts about how new account allocations will work? Could it be potentially dynamic such that CPI's could allocate many MB's of account data if needed? cc @brooksprumo

Reply:
maybe if the TX specifies the expected allocations in the TX